google-cloud-spanner/lib/google/cloud/spanner/pool.rb

```diff
@@ -59,31 +61,40 @@ def with_session
         end
 
         def checkout_session
+          action = nil
           @mutex.synchronize do
             loop do
+              fail ClientClosedError if @closed
+
               read_session = session_queue.shift
               return read_session if read_session
               write_transaction = transaction_queue.shift
               return write_transaction.session if write_transaction
 
-              return new_session! if can_allocate_more_sessions?
+              if can_allocate_more_sessions?
```

Comment:
this can cause you to go over limit. Consider if max sessions is 1 and currently there is no session. First thread enters this block and decides it can create a session. While it is creating, second thread can enter this block and also decide it can create one since there still are 0 sessions. So it will also go and create 1 more session. To avoid this you need to keep track of how many sessions are under creation right now and also factor this while calculating if you can allocate more sessions.